kv/kvnemesis: TestApplier failed

[(kv/kvnemesis).TestApplier failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2206464&tab=buildLog) on [master@991a05367d38ca32c1f6584e56143ac65270afc8](https://github.com/cockroachdb/cockroach/commits/991a05367d38ca32c1f6584e56143ac65270afc8):

```
  github.com/cockroachdb/cockroach/pkg/server.(*Node).RangeFeed()
      /go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:986 +0x78
  github.com/cockroachdb/cockroach/pkg/rpc.internalClientAdapter.RangeFeed.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/rpc/context.go:595 +0xf8

Previous write at 0x00c0015341d0 by goroutine 235:
  github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).RangeFeed.func2()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_rangefeed.go:214 +0x207
  github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed.(*registration).maybeRunCatchupScan()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed/registry.go:286 +0xa1
  github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed.(*registration).outputLoop()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed/registry.go:235 +0x53
  github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed.(*registration).runOutputLoop()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed/registry.go:271 +0x122
  github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed.(*Processor).run.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed/processor.go:240 +0x7e
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:347 +0x162

Goroutine 295 (running) created at:
  github.com/cockroachdb/cockroach/pkg/rpc.internalClientAdapter.RangeFeed()
      /go/src/github.com/cockroachdb/cockroach/pkg/rpc/context.go:593 +0x134
  github.com/cockroachdb/cockroach/pkg/rpc.(*internalClientAdapter).RangeFeed()
      <autogenerated>:1 +0xd9
  github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).singleRangeFeed()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go:257 +0x5b1
  github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).partialRangeFeed()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go:150 +0x3d6
  github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).RangeFeed.func1.1()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go:69 +0x7c
  github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.GoCtx.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:166 +0x4e
  golang.org/x/sync/errgroup.(*Group).Go.func1()
      /go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:57 +0x71

Goroutine 235 (finished) created at:
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:342 +0x14a
  github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed.(*Processor).run()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed/processor.go:246 +0x15f6
  github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed.(*Processor).Start.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed/processor.go:177 +0x67
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:347 +0x162
==================
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestApplier657779867
--- FAIL: TestApplier (1.15s)
    test_log_scope.go:154: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestApplier657779867
    test_log_scope.go:63: use -show-logs to present logs inline
    testing.go:853: race detected during execution of test

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-race -parallel=2

```
make stressrace TESTS=TestApplier PKG=./pkg/kv/kvnemesis TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestApplier.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(kv/kvnemesis).TestApplier failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2215082&tab=buildLog) on [master@b0ded83998c7b3b71d433dd93ad35a0fed561d3b](https://github.com/cockroachdb/cockroach/commits/b0ded83998c7b3b71d433dd93ad35a0fed561d3b):

```
  github.com/cockroachdb/cockroach/pkg/server.(*Node).RangeFeed()
      /go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:986 +0x78
  github.com/cockroachdb/cockroach/pkg/rpc.internalClientAdapter.RangeFeed.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/rpc/context.go:595 +0xf8

Previous write at 0x00c000ef2040 by goroutine 102:
  github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).RangeFeed.func2()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_rangefeed.go:214 +0x207
  github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed.(*registration).maybeRunCatchupScan()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed/registry.go:286 +0xa1
  github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed.(*registration).outputLoop()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed/registry.go:235 +0x53
  github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed.(*registration).runOutputLoop()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed/registry.go:271 +0x122
  github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed.(*Processor).run.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed/processor.go:240 +0x7e
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:347 +0x162

Goroutine 221 (running) created at:
  github.com/cockroachdb/cockroach/pkg/rpc.internalClientAdapter.RangeFeed()
      /go/src/github.com/cockroachdb/cockroach/pkg/rpc/context.go:593 +0x134
  github.com/cockroachdb/cockroach/pkg/rpc.(*internalClientAdapter).RangeFeed()
      <autogenerated>:1 +0xd9
  github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).singleRangeFeed()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go:257 +0x5b1
  github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).partialRangeFeed()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go:150 +0x3d6
  github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).RangeFeed.func1.1()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go:69 +0x7c
  github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.GoCtx.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:166 +0x4e
  golang.org/x/sync/errgroup.(*Group).Go.func1()
      /go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:57 +0x71

Goroutine 102 (running) created at:
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:342 +0x14a
  github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed.(*Processor).run()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed/processor.go:246 +0x15f6
  github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed.(*Processor).Start.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed/processor.go:177 +0x67
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:347 +0x162
==================
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestApplier312852580
--- FAIL: TestApplier (0.94s)
    test_log_scope.go:154: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestApplier312852580
    test_log_scope.go:63: use -show-logs to present logs inline
    testing.go:853: race detected during execution of test

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-race -parallel=2

```
make stressrace TESTS=TestApplier PKG=./pkg/kv/kvnemesis TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestApplier.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/kv/kvserver/replica_rangefeed.go
+++ b/pkg/kv/kvserver/replica_rangefeed.go
@@ -13,6 +13,7 @@ package kvserver
 import (
 	"context"
 	"fmt"
+	"sync"
 	"time"
 
 	"github.com/cockroachdb/cockroach/pkg/base"
@@ -170,14 +171,19 @@ func (r *Replica) RangeFeed(
 		if err := lim.Begin(ctx); err != nil {
 			return roachpb.NewError(err)
 		}
-		// Finish the iterator limit, but only if we exit before
-		// creating the iterator itself.
-		iterSemRelease = lim.Finish
-		defer func() {
-			if iterSemRelease != nil {
-				iterSemRelease()
-			}
-		}()
+		// Finish the iterator limit if we exit before the iterator finishes.
+		// The release function will be hooked into the Close method on the
+		// iterator below. The sync.Once prevents any races between exiting early
+		// from this call and finishing the catchup scan underneath the
+		// rangefeed.Processor. We need to release here in case we fail to
+		// register the processor, or, more perniciously, in the case where the
+		// processor gets registered by shut down before starting the catchup
+		// scan.
+		var iterSemReleaseOnce sync.Once
+		iterSemRelease = func() {
+			iterSemReleaseOnce.Do(lim.Finish)
+		}
+		defer iterSemRelease()
 	}
 
 	// Lock the raftMu, then register the stream as a new rangefeed registration.
@@ -210,8 +216,6 @@ func (r *Replica) RangeFeed(
 				SimpleIterator: innerIter,
 				close:          iterSemRelease,
 			}
-			// Responsibility for releasing the semaphore now passes to the iterator.
-			iterSemRelease = nil
 			return catchUpIter
 		}
 	}

